crates/red_knot_python_semantic/src/semantic_index/definition.rs

```diff
@@ -1,57 +1,133 @@
-use crate::semantic_index::ast_ids::{
-    ScopedClassId, ScopedExpressionId, ScopedFunctionId, ScopedStatementId,
-};
+use ruff_db::parsed::ParsedModule;
+use ruff_db::vfs::VfsFile;
+use ruff_python_ast as ast;
 
-#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
-pub enum Definition {
-    Import(ImportDefinition),
-    ImportFrom(ImportFromDefinition),
-    ClassDef(ScopedClassId),
-    FunctionDef(ScopedFunctionId),
-    Assignment(ScopedStatementId),
-    AnnotatedAssignment(ScopedStatementId),
-    NamedExpr(ScopedExpressionId),
-    /// represents the implicit initial definition of every name as "unbound"
-    Unbound,
-    // TODO with statements, except handlers, function args...
+use crate::ast_node_ref::AstNodeRef;
+use crate::node_key::NodeKey;
+use crate::semantic_index::symbol::{FileScopeId, ScopedSymbolId};
+
+#[salsa::tracked]
+pub struct Definition<'db> {
+    /// The file in which the definition is defined.
+    #[id]
+    pub(super) file: VfsFile,
+
+    /// The scope in which the definition is defined.
+    #[id]
+    pub(crate) scope: FileScopeId,
+
+    /// The id of the corresponding symbol. Mainly used as ID.
+    #[id]
+    symbol_id: ScopedSymbolId,
+
+    #[no_eq]
+    #[return_ref]
+    pub(crate) node: DefinitionKind,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub(crate) enum DefinitionNodeRef<'a> {
+    Alias(&'a ast::Alias),
+    Function(&'a ast::StmtFunctionDef),
+    Class(&'a ast::StmtClassDef),
+    NamedExpression(&'a ast::ExprNamed),
+    Target(&'a ast::Expr),
```

Comment:
Both the `Alias` and `Target` variants here will require walking up the AST to parents of the referenced node in order to get the necessary context to understand/resolve the definition. For `Alias` because we need info from the enclosing import node, and for `Target` because we also need to look at the targets of the assignment statement, not just the assigned (RHS) expression.

Is this going to be a problem? Do we have an efficient way of walking up the AST?

Reply:
I think this is not currently a problem because in the current code we are still only doing per-scope inference, so we always reach the alias in the context of visiting the import statement. But I think it will be a problem when we actually try to do inference on a Definition in isolation, and the ability to do that is the whole point of Definition being an ingredient.

Reply:
Yeah, I think we may want to change the representation. The good thing is, we very easily can. There's no constraint that a variant can only contain a single node. We can go back to storing the `ModuleName` and `Identifier` of the import or store the `Import` stmt with the alias's index. For now, what we have here is sufficient and is why I designed it this way.

Note: There might be some complication in how we build the map from `node` to `definition` because today `node` and the `definition` are kind of the same. Changing the granularity would e.g. mean that we map from `alias` to a definition storing a statement. But again, I don't think this breaks the design, it just complicates things a little more.